dali/kernels/imgproc/paste/paste_gpu.h

```diff
@@ -0,0 +1,247 @@
+
+  for (int y = threadIdx.y + block.start.y; y < block.end.y; y += blockDim.y) {
+    while (y >= my_grid_cells[grid_y * sample.cell_counts[1]].cell_end[0]) grid_y++;
+    int grid_x = 0;
+    for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
+      while (x >= my_grid_cells[grid_y * sample.cell_counts[1] + grid_x].cell_end[1]) grid_x++;
+      const GridCellGPU<InputType, ndims> *cell =
+              &my_grid_cells[grid_y * sample.cell_counts[1] + grid_x];
+      if (cell->in == nullptr) {
+          out[y * sample.out_pitch[1] + x] = 0;
+      } else {
+        out[y * sample.out_pitch[1] + x] = ConvertSat<OutputType>(
+                cell->in[(y - cell->cell_start[0] + cell->in_anchor[0]) * cell->in_pitch[1]
+                         + (x - cell->cell_start[1] + cell->in_anchor[1])]);
+      }
+    }
+  }
+}
+
+}  // namespace paste
+
+template <typename OutputType, typename InputType, int ndims>
+class PasteGPU {
+ private:
+  static constexpr size_t spatial_dims = ndims - 1;
+  using BlockDesc = kernels::BlockDesc<spatial_dims>;
+  using SampleDesc = paste::SampleDescriptorGPU<OutputType, InputType, spatial_dims>;
+  using GridCellDesc = paste::GridCellGPU<InputType, spatial_dims>;
+
+  std::vector<SampleDesc> sample_descriptors_;
+  std::vector<GridCellDesc> grid_cell_descriptors_;
+
+ public:
+  BlockSetup<spatial_dims, -1 /* No channel dimension, only spatial */> block_setup_;
+
+  KernelRequirements Setup(
+      KernelContext &context,
+      const InListGPU<InputType, ndims> &in,
+      const std::vector<paste::MultiPasteSampleInput<spatial_dims>> &samples,
+      const std::vector<paste::GridCellInput<spatial_dims>> &grid_cells,
+      TensorListShape<ndims> out_shape) {
+    DALI_ENFORCE([=]() -> bool {
+      auto ref_nchannels = in.shape[0][ndims - 1];
+      for (int i = 0; i < in.num_samples(); i++) {
+        if (in.shape[i][ndims - 1] != ref_nchannels) {
+          return false;
+        }
+      }
+      return true;
+    }(), "Number of channels for every image in batch must be equal");
+
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    TensorListShape<spatial_dims> flattened_shape(paste::FlattenChannels<ndims>(out_shape));
+    block_setup_.SetupBlocks(flattened_shape, true);
+    sample_descriptors_.resize(samples.size());
+    grid_cell_descriptors_.resize(grid_cells.size());
+    se.add<SampleDesc>(AllocType::GPU, samples.size());
+    se.add<GridCellDesc>(AllocType::GPU, grid_cells.size());
+    se.add<BlockDesc>(AllocType::GPU, block_setup_.Blocks().size());
+    // req.output_shapes = {in.shape}; Once again, this is determined by operator
+    req.scratch_sizes = se.sizes;
+    return req;
+  }
+
+
+  void Run(
+      KernelContext &context,
+      const OutListGPU<OutputType, ndims> &out, const InListGPU<InputType, ndims> &in,
+      const std::vector<paste::MultiPasteSampleInput<spatial_dims>> &samples,
+      const std::vector<paste::GridCellInput<spatial_dims>> &grid) {
```

Comment:
This isn't a good design - preparing the grid cells is by far the most complicated part of the whole affair - and it's left for the caller to provide. This is tied closely to the particular implementation and should either be a part of this kernel or a utility function shipped along the kernel - in any case, the algorithm of preparing the grid cells should be opaque - but see my other comments if that's indeed required.

Reply:
Moved patch generation to the kernel